### PR TITLE
 Introduce Grid2D as base class for 2D grids

### PR DESCRIPTION
**Pull request by kdaun**
This is a first step towards generalizing submaps (see [ongoing RFC discussion](https://github.com/googlecartographer/rfcs/pull/30)).
Although this PR already introduces a correspondence cost grid, the grid is still used as probability grid.

### CODE REVIEW

File: cartographer/mapping/2d/grid_2d.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+#ifndef CARTOGRAPHER_MAPPING_2D_GRID_2D_H_
+#define CARTOGRAPHER_MAPPING_2D_GRID_2D_H_
+
+#include <vector>
+
+#include "cartographer/mapping/2d/map_limits.h"
+#include "cartographer/mapping/2d/proto/grid_2d.pb.h"
+#include "cartographer/mapping/2d/xy_index.h"
```
Comment (pifon2a) [inline cartographer/mapping/2d/grid_2d.h]: Probably, you can move `xy_index.h` include to .cc file.

File: cartographer/mapping/2d/grid_2d.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace cartographer {
+namespace mapping {
+
+// Represents a 2D grid.
```
Comment (pifon2a) [inline cartographer/mapping/2d/grid_2d.h]: nit: this comment is not very useful. I would remove it.

File: cartographer/mapping/2d/grid_2d.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
+#include "cartographer/mapping/2d/grid_2d.h"
+
+#include <limits>
```
Comment (pifon2a) [inline cartographer/mapping/2d/grid_2d.cc]: is it used?

File: cartographer/mapping/2d/probability_grid.cc [hunk 4]
```diff
@@ -81,69 +69,16 @@ bool ProbabilityGrid::ApplyLookupTable(const Eigen::Array2i& cell_index,
 // Returns the probability of the cell with 'cell_index'.
 float ProbabilityGrid::GetProbability(const Eigen::Array2i& cell_index) const {
   if (limits_.Contains(cell_index)) {
-    return ValueToProbability(cells_[ToFlatIndex(cell_index, limits_)]);
+    return ValueToProbability(
```
Comment (pifon2a) [inline cartographer/mapping/2d/probability_grid.cc]: nit: you can swap these two statements, to reduce the number of lines:
```
if (!limits_.Contains(cell_index)) return kMinProbability;                       
return ValueToProbability(                                                       
    correspondence_cost_cells_[ToFlatIndex(cell_index, limits_)]); 
```

File: cartographer/mapping/2d/grid_2d.cc [hunk 5]
```diff
@@ -0,0 +1,102 @@
+#include "cartographer/mapping/2d/grid_2d.h"
```
Comment (gaschler) [inline cartographer/mapping/2d/grid_2d.cc]: add license

Comment (kdaun) [reply cartographer/mapping/2d/grid_2d.cc]: Added in https://github.com/googlecartographer/cartographer/pull/1046/commits/a9513a6c14bd7ae84093c90a7cfb33d23de97dff

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 4.

**Comment by wally-the-cartographer** [comment]
Refusing to merge, since the combined checks state is failure

File: cartographer/mapping/2d/grid_2d.cc [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+#include "cartographer/mapping/probability_values.h"
+
+namespace cartographer {
+namespace mapping {
+namespace {
+
+// Converts a 'cell_index' into an index into 'cells_'.
+int ToFlatIndex(const Eigen::Array2i& cell_index, const MapLimits& limits) {
```
Comment (pifon2a) [inline cartographer/mapping/2d/grid_2d.cc]: move it to Grid2D protected section. Currently it is in both grid_2d and in probability_grid_2d.

**Pull request merged**